Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,297 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+
+namespace Mirror
+{
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public abstract class SyncDictionary<B, T> : IDictionary<B, T>, SyncObject
+    {
+        public delegate void SyncDictionaryChanged(Operation op, B key, T item);
+
+        readonly Dictionary<B, T> m_Objects;
+
+        public int Count => m_Objects.Count;
+        public bool IsReadOnly { get; private set; }
+        public event SyncDictionaryChanged Callback;
+
+        public enum Operation : byte
+        {
+            OP_ADD,
+            OP_CLEAR,
+            OP_REMOVE,
+            OP_SET,
+            OP_DIRTY
+        }
+
+        struct Change
+        {
+            internal Operation operation;
+            internal B key;
+            internal T item;
+        }
+
+        readonly List<Change> Changes = new List<Change>();
+        // how many changes we need to ignore
+        // this is needed because when we initialize the list,
+        // we might later receive changes that have already been applied
+        // so we need to skip them
+        int changesAhead = 0;
+
+        protected virtual void SerializeKey(NetworkWriter writer, B item) {}
+        protected virtual void SerializeItem(NetworkWriter writer, T item) {}
+        protected virtual B DeserializeKey(NetworkReader reader) => default;
+        protected virtual T DeserializeItem(NetworkReader reader) => default;
+
+        public bool IsDirty => Changes.Count > 0;
+
+        public ICollection<B> Keys => m_Objects.Keys;
+
+        public ICollection<T> Values => m_Objects.Values;
+
+        // throw away all the changes
+        // this should be called after a successfull sync
+        public void Flush() => Changes.Clear();
```

Comment:
why is this public? shouldn't it be internal or similar? what if a user calls Flush(), isn't that a problem?

Reply:
It is required to implement IDictionary